In [1]:
!pip install frame-semantic-transformer

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl.metadata (1.8 kB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinu

In [2]:
!pip install spacy

In [3]:
!python -m spacy download en_core_web_lg

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 587.7/587.7 MB 2.3 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [4]:
import spacy

nlp = spacy.load('en_core_web_lg')

In [5]:
!pip install beautifulsoup4 html2text

import requests
from bs4 import BeautifulSoup
import html2text

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 kB 5.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for html2text: filename=html2text-2024.2.26-py3-none-any.whl size=33111 sha256=0edd40264b37a9d8d090df09423ebab28984405fecefa5d0162e08a70ac3a712
  Stored in directory: /root/.cache/pip/wheels/f3/96/6d/a7eba8f80d31cbd188a2787b81514d82fc5ae6943c44777659
Successfully built html2text


In [6]:
url = 'https://gutenberg.org/cache/epub/15701/pg15701-images.html'
extracted_html = requests.get(url)

if extracted_html.status_code == 200:
    extracted_content = extracted_html.text

In [8]:
soup = BeautifulSoup(extracted_content, 'html.parser')
elems = soup.find_all(["h2", "p"])

raw_sections = [""]
for e in elems:

    if e.name == "p":

        sub_elems = list(e.children)

        for s in sub_elems:

            if s.name is None:
                raw_sections[-1] += s

    else:
        raw_sections.append("")

output_sections = []
for r in raw_sections:

    if len(r) > 500:
        print(r[:500])
        print("_____________________")

        output_sections.append(r)

# 'output_sections' for final texts

My dear Lord,—The paper which I take the
liberty of sending to your Grace was, for the
greater part, written during the last session. A few
days after the prorogation some few observations were
added. I was, however, resolved to let it lie by me
for a considerable time, that, on viewing the matter
at a proper distance, and when the sharpness of recent
impressions had been worn off, I might be better
able to form a just estimate of the value of my
first opinions.I have just now read it o
_____________________
Approaching towards the close of a long period
of public service, it is natural I should be
desirous to stand well (I hope I do stand tolerably
well) with that public which, with whatever fortune,
I have endeavored faithfully and zealously to serve.I am also not a little anxious for some place in the
estimation of the two persons to whom I address this
paper. I have always acted with them, and with
those whom they represent. To my knowledge, I
have not deviated, no, not in the minu

In this last two cells of code, after installing and importing all the necessary libraries, I import the specific text through 'requests' an 'beautiful soup'. The for loop in the cell above simply looks across all the content extracted from the project guttenberg webpage and finds instances of text 'p', which are indexed as actual text and not appendices, footnotes, and copyright information. This is indicated when 's' eqauls NONE. Therefore this code gives back only the actual written text of the auhtor.   

In [11]:
cut_text_B_8= output_sections[:2]
#cut_text_B_8

output_sections is numbered by sections of the text. Because the entirety of output_sections comprises of multiple letters of and phamplets of Edmund Burke's life, I wanted to break it up into each text so that I could label them and know later down the line which sentence belongs to which text. In this case 'B_8_cut_text' stands for a particular letter of Burkes. The next letter is B_9_cut_text and so on. However this step isn't neccessary and one could just join output sections into one string and work from there.

In [13]:
B_8_raws = "\n".join(cut_text_B_8)
B_8_raw= B_8_raws.lower()
#B_8_raw

In [15]:
B_8_raws = B_8_raw.replace("\\", "")

B_8_raw = B_8_raw.replace("\n", " ")
B_8_raw = B_8_raw.replace("\r", " ")
B_8_raw = B_8_raw.replace("_", "")
B_8_raw = B_8_raw.replace("  ", " ")
#B_8_raw

these two cells joins the text into one string and then '.replaces' and unwanted characters with empty spaces

In [21]:
doc = nlp(B_8_raw)

sentences_B_8 = []

for sent in doc.sents:
  sentences_B_8.append(sent.text)

#sentences_B_8

In [19]:
len(sentences_B_8)

475

Here I then run the 'en_core_web_sum_lg' large language model from Spacy to split the text per sentence which gives a accurate list of sentences, from the text. As we can see this list of sentences (sentences_B_8) has 475 sentences within it.

In [22]:
!pip install nltk
import nltk
nltk.download('punkt')
nltk.download('wordnet')

from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


In [23]:
base_targets = ['despotism', 'democracy', 'republican', 'republics', 'government', 'monarchy', 'burthens', 'magistracy', 'monarchies', 'despotic', 'political', 'politics', 'politic', 'state', 'tyranny', 'oppression', 'oppressed', 'hereditary', 'nation', 'anarchy', 'liberty', 'freedom', 'revolution', 'patriots', 'patriotism', 'equality', 'rights', 'domination', 'slavery', 'licentiousness', 'aristocracy', 'executive', 'legislative', 'civil', 'laws', 'constitution', 'polity', 'discontents', 'subversion']
target_lemmas = [lemmatizer.lemmatize(word) for word in base_targets]


target_sentences_B_8 = []
for sent in sentences_B_8:
  tokenized = word_tokenize(sent)
  lemmas = [lemmatizer.lemmatize(word) for word in tokenized]
  if any(lemma in target_lemmas for lemma in lemmas):
      target_sentences_B_8.append(sent)
  else:
    continue
#target_sentences_B_8
len(target_sentences_B_8)


115

In this cell I filter out sentences by checking to see if each sentence has any words found in target_lemmas. Only sentences that do are appened to 'target_sentences_B_8'. As we can see this now gives us a list of 115 sentences which cut about three quartes of the original text.

In [ ]:
from frame_semantic_transformer import FrameSemanticTransformer
frame_transformer = FrameSemanticTransformer()

In [ ]:
result_B_8 = frame_transformer.detect_frames_bulk(target_sentences_B_8)

for i in result_B_8:
  print(f"Results found in: {i.sentence}")
  for frame in i.frames:
    print(f"FRAME: {frame.name}")
    for element in frame.frame_elements:
        print(f"{element.name}: {element.text}")

This cell now runs inputs each sentence into the frame semantic transformer and holds the data in result_B_8. WARNING, only run this cell if you actually want to get the results. It can take a long time to complete and is very computationally intensive. In this case it won't take too long maybe 10 minutes.

If you do decide to run it make sure to save the results as a pickle file so that you don't have to run it again.